The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6065942745/job/16455980640) [(plain)](https://github.com/rust-lang/rust/commit/84b158ae4c27ac08db5811e68303a4577d4b8232/checks/16455980640/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=saethlin
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_a70ad8b9-9e03-4a5e-bfd2-58cbdc433b98
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=better-u32-encoding
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_a70ad8b9-9e03-4a5e-bfd2-58cbdc433b98
GITHUB_REF=refs/pull/110050/merge
GITHUB_REF_NAME=110050/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=9a83583ba8cfc2ec34bed8d0ec528c2aaf5c2657
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_a70ad8b9-9e03-4a5e-bfd2-58cbdc433b98
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_a70ad8b9-9e03-4a5e-bfd2-58cbdc433b98
GITHUB_TRIGGERING_ACTOR=saethlin
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/110050/merge
GITHUB_WORKFLOW_SHA=9a83583ba8cfc2ec34bed8d0ec528c2aaf5c2657
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
........................................................................................   616/15520
........................................................................................   704/15520
....................................................................F...................   792/15520
........................................................................................   880/15520
.................F...................................................F..................   968/15520
..................................................................................i.....  1144/15520
........................................................................................  1232/15520
........................................................................................  1320/15520
........................................................................................  1408/15520
---
........................................................................................  3344/15520
........................................................................................  3432/15520
........................................................................................  3520/15520
..................................................................................i.....  3608/15520
..................................F.i.F.F.FF..F..F...F..................................  3696/15520
......................................................................................ii  3872/15520
iii.....................................................................................  3960/15520
......................................................................................F.  4048/15520
........................................................................................  4136/15520
---
...................i....................................................................  4840/15520
....................................i...................................................  4928/15520
........................................................................................  5016/15520
........................................................................................  5104/15520
................F..F....................................................................  5192/15520
.............i..........................................................................  5368/15520
........................................................................................  5456/15520
........................................................................................  5544/15520
........................................................................................  5632/15520
---
7 error[E0107]: missing generics for struct `Local`
8   --> $DIR/issue-109768.rs:7:9
9    |

20 LL | impl<T> Local<T, U> {
22 
+ 9
+ 2
+ 9
---
To only update this specific test, also pass `--test-args associated-inherent-types/issue-109768.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/associated-inherent-types/issue-109768.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-inherent-types/issue-109768/issue-109768.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-inherent-types/issue-109768" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-inherent-types/issue-109768/auxiliary"
--- stderr -------------------------------
9
2
9
---
2
error[E0107]: missing generics for struct `Local`
##[error]  --> /checkout/tests/ui/associated-inherent-types/issue-109768.rs:7:9
   |
LL | impl<T> Local { //~ ERROR missing generics for struct `Local`
   |         ^^^^^ expected 2 generic arguments
   |
note: struct defined here, with 2 generic parameters: `T`, `U`
   |
   |
LL | struct Local<T, U>(T, U);
   |        ^^^^^ -  -
help: add missing generic arguments
   |
LL | impl<T> Local<T, U> { //~ ERROR missing generics for struct `Local`

9
2
9
---
2
error[E0658]: inherent associated types are unstable
##[error]  --> /checkout/tests/ui/associated-inherent-types/issue-109768.rs:8:5
   |
LL |     type AssocType3 = T; //~ ERROR inherent associated types are unstable
   |
   = note: see issue #8995 <https://github.com/rust-lang/rust/issues/8995> for more information
   = help: add `#![feature(inherent_associated_types)]` to the crate attributes to enable

---
To only update this specific test, also pass `--test-args associated-type-bounds/traits-assoc-type-macros.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/associated-type-bounds/traits-assoc-type-macros.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-type-bounds/traits-assoc-type-macros/traits-assoc-type-macros.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-type-bounds/traits-assoc-type-macros" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-type-bounds/traits-assoc-type-macros/auxiliary"
--- stderr -------------------------------
9
2
9
---
1 error: implementation of `FnOnce` is not general enough
2   --> $DIR/drop-tracking-unresolved-typeck-results.rs:99:5
3    |

58    = note: `fn(&'0 ()) -> std::future::Ready<&'0 ()> {std::future::ready::<&'0 ()>}` must implement `FnOnce<(&'1 (),)>`, for any two lifetimes `'0` and `'1`...
59    = note: ...but it actually implements `FnOnce<(&(),)>`
+ 9
+ 2
+ 9
+ 2
---
To only update this specific test, also pass `--test-args async-await/drop-tracking-unresolved-typeck-results.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/async-await/drop-tracking-unresolved-typeck-results.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/drop-tracking-unresolved-typeck-results/drop-tracking-unresolved-typeck-results.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/drop-tracking-unresolved-typeck-results" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/drop-tracking-unresolved-typeck-results/auxiliary" "-Zdrop-tracking" "--edition=2021"
--- stderr -------------------------------
9
2
9
---
2
error: implementation of `FnOnce` is not general enough
##[error]  --> /checkout/tests/ui/async-await/drop-tracking-unresolved-typeck-results.rs:99:5
   |
LL | /     send(async {
LL | |         //~^ ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         Next(&Buffered(Map(Empty(PhantomData), ready::<&()>), FuturesOrdered(PhantomData), 0)).await
LL | |     });
   | |______^ implementation of `FnOnce` is not general enough
   |
   = note: `fn(&'0 ()) -> std::future::Ready<&'0 ()> {std::future::ready::<&'0 ()>}` must implement `FnOnce<(&'1 (),)>`, for any two lifetimes `'0` and `'1`...
   = note: ...but it actually implements `FnOnce<(&(),)>`
error: implementation of `FnOnce` is not general enough
##[error]  --> /checkout/tests/ui/async-await/drop-tracking-unresolved-typeck-results.rs:99:5
   |
LL | /     send(async {
LL | /     send(async {
LL | |         //~^ ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         Next(&Buffered(Map(Empty(PhantomData), ready::<&()>), FuturesOrdered(PhantomData), 0)).await
LL | |     });
   | |______^ implementation of `FnOnce` is not general enough
   |
   = note: `fn(&'0 ()) -> std::future::Ready<&'0 ()> {std::future::ready::<&'0 ()>}` must implement `FnOnce<(&'1 (),)>`, for any two lifetimes `'0` and `'1`...
   = note: ...but it actually implements `FnOnce<(&(),)>`
error: implementation of `FnOnce` is not general enough
##[error]  --> /checkout/tests/ui/async-await/drop-tracking-unresolved-typeck-results.rs:99:5
   |
LL | /     send(async {
LL | /     send(async {
LL | |         //~^ ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         Next(&Buffered(Map(Empty(PhantomData), ready::<&()>), FuturesOrdered(PhantomData), 0)).await
LL | |     });
   | |______^ implementation of `FnOnce` is not general enough
   |
   = note: `fn(&'0 ()) -> std::future::Ready<&'0 ()> {std::future::ready::<&'0 ()>}` must implement `FnOnce<(&'1 (),)>`, for any two lifetimes `'0` and `'1`...
   = note: ...but it actually implements `FnOnce<(&(),)>`
error: implementation of `FnOnce` is not general enough
##[error]  --> /checkout/tests/ui/async-await/drop-tracking-unresolved-typeck-results.rs:99:5
   |
LL | /     send(async {
LL | /     send(async {
LL | |         //~^ ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         //~| ERROR implementation of `FnOnce` is not general enough
LL | |         Next(&Buffered(Map(Empty(PhantomData), ready::<&()>), FuturesOrdered(PhantomData), 0)).await
LL | |     });
   | |______^ implementation of `FnOnce` is not general enough
   |
   = note: `fn(&'0 ()) -> std::future::Ready<&'0 ()> {std::future::ready::<&'0 ()>}` must implement `FnOnce<(&'1 (),)>`, for any two lifetimes `'0` and `'1`...
   = note: ...but it actually implements `FnOnce<(&(),)>`
9
2
9
2
---
+ 9
+ 2
+ 9
+ 2
1 error[E0277]: the trait bound `Option<&str>: AsRef<Path>` is not satisfied
3    |

9 note: required by a bound in `File::open`
10   --> $SRC_DIR/std/src/fs.rs:LL:COL
---
To only update this specific test, also pass `--test-args async-await/issue-72442.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/async-await/issue-72442.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/issue-72442/issue-72442.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/issue-72442" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/issue-72442/auxiliary" "--edition=2018"
--- stderr -------------------------------
9
2
9
---
9
2
9
2
error[E0277]: the trait bound `Option<&str>: AsRef<Path>` is not satisfied
   |
LL |             let mut f = File::open(path.to_str())?;
LL |             let mut f = File::open(path.to_str())?;
   |                         ---------- ^^^^^^^^^^^^^ the trait `AsRef<Path>` is not implemented for `Option<&str>`
   |                         required by a bound introduced by this call
   |
note: required by a bound in `File::open`
  --> /rustc/FAKE_PREFIX/library/std/src/fs.rs:364:5
---
To only update this specific test, also pass `--test-args async-await/issues/issue-64964.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/async-await/issues/issue-64964.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/issues/issue-64964/issue-64964.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/issues/issue-64964" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/issues/issue-64964/auxiliary" "-Z" "query-dep-graph" "--edition=2018"
--- stderr -------------------------------
9
2
9
---
1 error[E0308]: mismatched types
2   --> $DIR/bad-subst-const-kind.rs:8:31
3    |

4 LL | impl<const N: u64> Q for [u8; N] {
5    |                               ^ expected `usize`, found `u64`
+ 9
+ 2
+ 9
+ 2
---
To only update this specific test, also pass `--test-args const-generics/bad-subst-const-kind.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/const-generics/bad-subst-const-kind.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/bad-subst-const-kind/bad-subst-const-kind.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/bad-subst-const-kind" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/bad-subst-const-kind/auxiliary"
--- stderr -------------------------------
9
2
9
---
2
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/const-generics/bad-subst-const-kind.rs:8:31
   |
LL | impl<const N: u64> Q for [u8; N] {
   |                               ^ expected `usize`, found `u64`
9
2
9
2
---
1 error: OK
2   --> $DIR/dep-graph-caller-callee.rs:21:5
3    |

4 LL |     #[rustc_then_this_would_need(typeck)]
6 
+ 9
+ 2
+ 9
---
+ 9
+ 2
+ 9
+ 2
7 error: no path from `x` to `typeck`
9    |


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-caller-callee/dep-graph-caller-callee.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args dep-graph/dep-graph-caller-callee.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/dep-graph/dep-graph-caller-callee.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-caller-callee/dep-graph-caller-callee.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-caller-callee" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-caller-callee/auxiliary" "-Z" "query-dep-graph"
--- stderr -------------------------------
9
2
9
---
2
error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-caller-callee.rs:21:5
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR OK

9
2
9
---
9
2
9
2
error: no path from `x` to `typeck`
   |
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR no path

error: aborting due to 2 previous errors
------------------------------------------

---
To only update this specific test, also pass `--test-args dep-graph/dep-graph-assoc-type-codegen.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/dep-graph/dep-graph-assoc-type-codegen.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-assoc-type-codegen/dep-graph-assoc-type-codegen.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-assoc-type-codegen" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-assoc-type-codegen/auxiliary" "-Z" "query-dep-graph"
--- stderr -------------------------------
9
2
9
---
2
error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-assoc-type-codegen.rs:29:5
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR OK

error: aborting due to previous error
------------------------------------------

---
1 error: OK
2   --> $DIR/dep-graph-trait-impl-two-traits.rs:32:5
3    |

4 LL |     #[rustc_then_this_would_need(typeck)]
6 
+ 9
+ 2
+ 9
---
+ 9
+ 2
+ 9
+ 2
7 error: no path from `x::<impl Foo for char>` to `typeck`
9    |


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl-two-traits/dep-graph-trait-impl-two-traits.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args dep-graph/dep-graph-trait-impl-two-traits.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/dep-graph/dep-graph-trait-impl-two-traits.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl-two-traits/dep-graph-trait-impl-two-traits.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl-two-traits" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl-two-traits/auxiliary" "-Z" "query-dep-graph"
--- stderr -------------------------------
9
2
9
---
2
error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-trait-impl-two-traits.rs:32:5
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR OK

9
2
9
---
9
2
9
2
error: no path from `x::<impl Foo for char>` to `typeck`
   |
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR no path

error: aborting due to 2 previous errors
------------------------------------------

---
+ 9
+ 2
+ 9
+ 2
1 error: no path from `WillChange` to `type_of`
3    |


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-struct-signature/dep-graph-struct-signature.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args dep-graph/dep-graph-struct-signature.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-struct-signature/dep-graph-struct-signature.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-struct-signature" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-struct-signature/auxiliary" "-Z" "query-dep-graph"
--- stderr -------------------------------
9
2
9
---
9
2
9
2
error: no path from `WillChange` to `type_of`
   |
   |
LL |     #[rustc_then_this_would_need(type_of)] //~ ERROR no path


error: no path from `WillChange` to `associated_item`
   |
   |
LL |     #[rustc_then_this_would_need(associated_item)] //~ ERROR no path


error: no path from `WillChange` to `trait_def`
   |
   |
LL |     #[rustc_then_this_would_need(trait_def)] //~ ERROR no path

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:36:5
   |
   |
LL |     #[rustc_then_this_would_need(fn_sig)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:37:5
   |
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:40:5
   |
   |
LL |     #[rustc_then_this_would_need(fn_sig)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:41:5
   |
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:46:5
   |
   |
LL |     #[rustc_then_this_would_need(type_of)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:53:5
   |
   |
LL |     #[rustc_then_this_would_need(type_of)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:61:9
   |
   |
LL |         #[rustc_then_this_would_need(type_of)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:63:9
   |
   |
LL |         #[rustc_then_this_would_need(type_of)] //~ ERROR OK


error: no path from `WillChange` to `type_of`
   |
   |
LL |     #[rustc_then_this_would_need(type_of)] //~ ERROR no path


error: no path from `WillChange` to `type_of`
   |
   |
LL |     #[rustc_then_this_would_need(type_of)] //~ ERROR no path


error: no path from `WillChange` to `fn_sig`
   |
   |
LL |     #[rustc_then_this_would_need(fn_sig)] //~ ERROR no path


error: no path from `WillChange` to `fn_sig`
   |
   |
LL |     #[rustc_then_this_would_need(fn_sig)] //~ ERROR no path from `WillChange`


error: no path from `WillChange` to `typeck`
   |
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR no path from `WillChange`

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:32:9
   |
   |
LL |         #[rustc_then_this_would_need(fn_sig)] //~ ERROR OK


error: no path from `WillChange` to `fn_sig`
   |
   |
LL |         #[rustc_then_this_would_need(fn_sig)] //~ ERROR no path

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:48:9
   |
   |
LL |         #[rustc_then_this_would_need(fn_sig)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:49:9
   |
   |
LL |         #[rustc_then_this_would_need(typeck)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:55:9
   |
   |
LL |         #[rustc_then_this_would_need(fn_sig)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-struct-signature.rs:56:9
   |
   |
LL |         #[rustc_then_this_would_need(typeck)] //~ ERROR OK

error: aborting due to 22 previous errors
------------------------------------------

---
1 error: OK
2   --> $DIR/dep-graph-trait-impl-two-traits-same-method.rs:33:5
3    |

4 LL |     #[rustc_then_this_would_need(typeck)]
6 
+ 9
+ 2
+ 9
---
+ 9
+ 2
+ 9
+ 2
7 error: no path from `x::<impl Foo for u32>` to `typeck`
9    |


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl-two-traits-same-method/dep-graph-trait-impl-two-traits-same-method.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args dep-graph/dep-graph-trait-impl-two-traits-same-method.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/dep-graph/dep-graph-trait-impl-two-traits-same-method.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl-two-traits-same-method/dep-graph-trait-impl-two-traits-same-method.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl-two-traits-same-method" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl-two-traits-same-method/auxiliary" "-Z" "query-dep-graph"
--- stderr -------------------------------
9
2
9
---
2
error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-trait-impl-two-traits-same-method.rs:33:5
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR OK

9
2
9
---
9
2
9
2
error: no path from `x::<impl Foo for u32>` to `typeck`
   |
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR no path

error: aborting due to 2 previous errors
------------------------------------------

---
1 error: OK
2   --> $DIR/dep-graph-trait-impl.rs:28:5
3    |

22 LL |     #[rustc_then_this_would_need(typeck)]
24 
+ 9
+ 2
+ 9
---
+ 9
+ 2
+ 9
+ 2
25 error: no path from `x::<impl Foo for char>` to `typeck`
27    |


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl/dep-graph-trait-impl.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args dep-graph/dep-graph-trait-impl.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/dep-graph/dep-graph-trait-impl.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl/dep-graph-trait-impl.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-trait-impl/auxiliary" "-Z" "query-dep-graph"
--- stderr -------------------------------
9
2
9
---
2
error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-trait-impl.rs:28:5
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-trait-impl.rs:33:5
   |
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-trait-impl.rs:38:5
   |
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-trait-impl.rs:43:5
   |
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR OK

9
2
9
---
9
2
9
2
error: no path from `x::<impl Foo for char>` to `typeck`
   |
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR no path

error: aborting due to 5 previous errors
------------------------------------------

---
To only update this specific test, also pass `--test-args dep-graph/dep-graph-variance-alias.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/dep-graph/dep-graph-variance-alias.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-variance-alias/dep-graph-variance-alias.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-variance-alias" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-variance-alias/auxiliary" "-Z" "query-dep-graph"
--- stderr -------------------------------
9
2
9
---
2
error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-variance-alias.rs:19:1
   |
LL | #[rustc_then_this_would_need(variances_of)] //~ ERROR OK

error: aborting due to previous error
------------------------------------------

---
+ 9
+ 2
+ 9
+ 2
1 error: no path from `TypeAlias` to `type_of`
3    |


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-type-alias/dep-graph-type-alias.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args dep-graph/dep-graph-type-alias.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/dep-graph/dep-graph-type-alias.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-type-alias/dep-graph-type-alias.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-type-alias" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dep-graph/dep-graph-type-alias/auxiliary" "-Z" "query-dep-graph"
--- stderr -------------------------------
9
2
9
---
9
2
9
2
error: no path from `TypeAlias` to `type_of`
   |
   |
LL | #[rustc_then_this_would_need(type_of)] //~ ERROR no path

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-type-alias.rs:20:5
   |
   |
LL |     #[rustc_then_this_would_need(type_of)] //~ ERROR OK


error: no path from `TypeAlias` to `type_of`
   |
   |
LL | #[rustc_then_this_would_need(type_of)] //~ ERROR no path

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-type-alias.rs:28:9
   |
   |
LL |         #[rustc_then_this_would_need(type_of)] //~ ERROR OK


error: no path from `TypeAlias` to `type_of`
   |
   |
LL | #[rustc_then_this_would_need(type_of)] //~ ERROR no path


error: no path from `TypeAlias` to `type_of`
   |
   |
LL | #[rustc_then_this_would_need(type_of)] //~ ERROR no path

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-type-alias.rs:49:1
   |
   |
LL | #[rustc_then_this_would_need(type_of)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-type-alias.rs:52:1
   |
   |
LL | #[rustc_then_this_would_need(fn_sig)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-type-alias.rs:53:1
   |
   |
LL | #[rustc_then_this_would_need(typeck)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-type-alias.rs:36:5
   |
   |
LL |     #[rustc_then_this_would_need(fn_sig)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-type-alias.rs:44:5
   |
   |
LL |     #[rustc_then_this_would_need(fn_sig)] //~ ERROR OK

error: OK
##[error]  --> /checkout/tests/ui/dep-graph/dep-graph-type-alias.rs:45:5
   |
   |
LL |     #[rustc_then_this_would_need(typeck)] //~ ERROR OK

error: aborting due to 12 previous errors
------------------------------------------

---
9 For more information about this error, try `rustc --explain E0282`.


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dyn-star/param-env-region-infer.current/param-env-region-infer.current.stderr
To only update this specific test, also pass `--test-args dyn-star/param-env-region-infer.rs`


error in revision `current`: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/dyn-star/param-env-region-infer.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "current" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dyn-star/param-env-region-infer.current/param-env-region-infer.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dyn-star/param-env-region-infer.current" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/dyn-star/param-env-region-infer.current/auxiliary"
--- stderr -------------------------------
9
2
9
---
+ 9
+ 2
+ 9
+ 2
1 error[E0277]: `[(); _]` is not a future
3    |


12    = note: [(); _] must be a future or must implement `IntoFuture` to be awaited
13    = note: required for `[(); _]` to implement `IntoFuture`
+ 9
+ 2
+ 9
+ 2
---
+ 9
+ 2
+ 9
+ 2
15 error[E0698]: type inside `async` block must be known in this context
17    |


72 LL |         let s = std::array::from_fn(|_| ()).await;
74 
+ 9
+ 2
+ 9
---
To only update this specific test, also pass `--test-args generator/unresolved-ct-var-drop-tracking.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/generator/unresolved-ct-var-drop-tracking.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generator/unresolved-ct-var-drop-tracking/unresolved-ct-var-drop-tracking.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generator/unresolved-ct-var-drop-tracking" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generator/unresolved-ct-var-drop-tracking/auxiliary" "--edition=2021" "-Zdrop-tracking"
--- stderr -------------------------------
9
2
9
---
9
2
9
2
error[E0277]: `[(); _]` is not a future
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 |                          ||
   |                 |                          ||
   |                 |                          |`[(); _]` is not a future
   |                 |                          help: remove the `.await`
   |                 this call returns `[(); _]`
   |
   = help: the trait `Future` is not implemented for `[(); _]`
   = note: [(); _] must be a future or must implement `IntoFuture` to be awaited
   = note: required for `[(); _]` to implement `IntoFuture`
9
2
9
2
---
9
2
9
2
error[E0698]: type inside `async` block must be known in this context
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 ^^^^^^^^^^^^^^^^^^^ cannot infer the value of const parameter `N` declared on the function `from_fn`
   |
note: the type is part of the `async` block because of this `await`
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;


error[E0698]: type inside `async` block must be known in this context
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 ^^^^^^^^^^^^^^^^^^^ cannot infer the value of const parameter `N` declared on the function `from_fn`
   |
note: the type is part of the `async` block because of this `await`
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;


error[E0698]: type inside `async` block must be known in this context
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 ^^^^^^^^^^^^^^^^^^^ cannot infer the value of const parameter `N` declared on the function `from_fn`
   |
note: the type is part of the `async` block because of this `await`
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;


error[E0698]: type inside `async` block must be known in this context
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 ^^^^^^^^^^^^^^^^^^^ cannot infer the value of const parameter `N` declared on the function `from_fn`
   |
note: the type is part of the `async` block because of this `await`
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;


error[E0698]: type inside `async` block must be known in this context
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 ^^^^^^^^^^^^^^^^^^^ cannot infer the value of const parameter `N` declared on the function `from_fn`
   |
note: the type is part of the `async` block because of this `await`
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;

9
2
9
---
+ 9
+ 2
+ 9
+ 2
1 error[E0277]: `[(); _]` is not a future
3    |


12    = note: [(); _] must be a future or must implement `IntoFuture` to be awaited
13    = note: required for `[(); _]` to implement `IntoFuture`
+ 9
+ 2
+ 9
+ 2
---
+ 9
+ 2
+ 9
+ 2
15 error[E0698]: type inside `async` block must be known in this context
17    |


72 LL |         let s = std::array::from_fn(|_| ()).await;
74 
+ 9
+ 2
+ 9
---
To only update this specific test, also pass `--test-args generator/unresolved-ct-var.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/generator/unresolved-ct-var.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generator/unresolved-ct-var/unresolved-ct-var.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generator/unresolved-ct-var" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generator/unresolved-ct-var/auxiliary" "--edition=2021"
--- stderr -------------------------------
9
2
9
---
9
2
9
2
error[E0277]: `[(); _]` is not a future
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 |                          ||
   |                 |                          ||
   |                 |                          |`[(); _]` is not a future
   |                 |                          help: remove the `.await`
   |                 this call returns `[(); _]`
   |
   = help: the trait `Future` is not implemented for `[(); _]`
   = note: [(); _] must be a future or must implement `IntoFuture` to be awaited
   = note: required for `[(); _]` to implement `IntoFuture`
9
2
9
2
---
9
2
9
2
error[E0698]: type inside `async` block must be known in this context
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 ^^^^^^^^^^^^^^^^^^^ cannot infer the value of const parameter `N` declared on the function `from_fn`
   |
note: the type is part of the `async` block because of this `await`
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;


error[E0698]: type inside `async` block must be known in this context
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 ^^^^^^^^^^^^^^^^^^^ cannot infer the value of const parameter `N` declared on the function `from_fn`
   |
note: the type is part of the `async` block because of this `await`
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;


error[E0698]: type inside `async` block must be known in this context
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 ^^^^^^^^^^^^^^^^^^^ cannot infer the value of const parameter `N` declared on the function `from_fn`
   |
note: the type is part of the `async` block because of this `await`
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;


error[E0698]: type inside `async` block must be known in this context
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 ^^^^^^^^^^^^^^^^^^^ cannot infer the value of const parameter `N` declared on the function `from_fn`
   |
note: the type is part of the `async` block because of this `await`
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;


error[E0698]: type inside `async` block must be known in this context
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;
   |                 ^^^^^^^^^^^^^^^^^^^ cannot infer the value of const parameter `N` declared on the function `from_fn`
   |
note: the type is part of the `async` block because of this `await`
   |
   |
LL |         let s = std::array::from_fn(|_| ()).await;

9
2
9
---
1 error[E0277]: the `?` operator can only be applied to values that implement `Try`
2   --> $DIR/issue-72766.rs:14:5
3    |

10 LL |     SadGirl {}.call().await?;
12 
+ 9
+ 2
+ 9
---
To only update this specific test, also pass `--test-args suggestions/issue-72766.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/suggestions/issue-72766.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/issue-72766/issue-72766.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/issue-72766" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/issue-72766/auxiliary" "--edition=2018"
--- stderr -------------------------------
9
2
9
---
2
error[E0277]: the `?` operator can only be applied to values that implement `Try`
##[error]  --> /checkout/tests/ui/suggestions/issue-72766.rs:14:5
   |
LL |     SadGirl {}.call()?; //~ ERROR: the `?` operator can only be applied to values
   |     ^^^^^^^^^^^^^^^^^^ the `?` operator cannot be applied to type `impl Future<Output = Result<(), ()>>`
   = help: the trait `Try` is not implemented for `impl Future<Output = Result<(), ()>>`
   = help: the trait `Try` is not implemented for `impl Future<Output = Result<(), ()>>`
help: consider `await`ing on the `Future`
   |
LL |     SadGirl {}.call().await?; //~ ERROR: the `?` operator can only be applied to values

9
2
9
---
To only update this specific test, also pass `--test-args traits/vtable/issue-91807.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/traits/vtable/issue-91807.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/vtable/issue-91807/issue-91807.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/vtable/issue-91807" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/vtable/issue-91807/auxiliary"
--- stderr -------------------------------
9
2
9
---
+ 9
+ 2
+ 9
+ 2
1 error[E0277]: `&'a T` is not an iterator
2   --> $DIR/hir-wf-check-erase-regions.rs:7:21

10 note: required by a bound in `Flatten`
11   --> $SRC_DIR/core/src/iter/adapters/flatten.rs:LL:COL
12 
---
+ 9
+ 2
+ 9
+ 2
13 error[E0277]: `&'a T` is not an iterator
14   --> $DIR/hir-wf-check-erase-regions.rs:10:27

22 note: required by a bound in `Flatten`
23   --> $SRC_DIR/core/src/iter/adapters/flatten.rs:LL:COL
24 
---
To only update this specific test, also pass `--test-args wf/hir-wf-check-erase-regions.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/wf/hir-wf-check-erase-regions.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/wf/hir-wf-check-erase-regions/hir-wf-check-erase-regions.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/wf/hir-wf-check-erase-regions" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/wf/hir-wf-check-erase-regions/auxiliary"
--- stderr -------------------------------
9
2
9
---
9
2
9
2
error[E0277]: `&'a T` is not an iterator
   |
   |
LL |     type IntoIter = std::iter::Flatten<std::slice::Iter<'a, T>>; //~ ERROR `&'a T` is not an iterator
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `&'a T` is not an iterator
   |
   = help: the trait `Iterator` is not implemented for `&'a T`
   = help: the trait `Iterator` is implemented for `&mut I`
   = note: required for `&'a T` to implement `IntoIterator`
note: required by a bound in `Flatten`
  --> /rustc/FAKE_PREFIX/library/core/src/iter/adapters/flatten.rs:157:1
9
2
9
2
---
9
2
9
2
error[E0277]: `&'a T` is not an iterator
   |
   |
LL |     fn into_iter(self) -> Self::IntoIter { //~ ERROR `&'a T` is not an iterator
   |                           ^^^^^^^^^^^^^^ `&'a T` is not an iterator
   |
   = help: the trait `Iterator` is not implemented for `&'a T`
   = help: the trait `Iterator` is implemented for `&mut I`
   = note: required for `&'a T` to implement `IntoIterator`
note: required by a bound in `Flatten`
  --> /rustc/FAKE_PREFIX/library/core/src/iter/adapters/flatten.rs:157:1
9
2
9
2
---
+ 9
+ 2
+ 9
+ 2
1 error[E0277]: the trait bound `Bar<'a, T>: Foo` is not satisfied
2   --> $DIR/hir-wf-canonicalized.rs:10:15


10 LL |     callback: Box<dyn Callback<dyn Callback<Bar<'a, T>>>>,
11    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Foo` is not implemented for `(dyn Callback<Bar<'a, T>, for<'b, 'c, 'd> Output = ()> + 'static)`
+ 9
+ 2
+ 9
+ 2
---
+ 9
+ 2
+ 9
+ 2
13 error[E0277]: the size for values of type `(dyn Callback<Bar<'a, T>, for<'b, 'c, 'd> Output = ()> + 'static)` cannot be known at compilation time
14   --> $DIR/hir-wf-canonicalized.rs:10:15


27 LL | struct Bar<'a, T: ?Sized> {
29 
+ 9
+ 2
+ 9
---
To only update this specific test, also pass `--test-args wf/hir-wf-canonicalized.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/wf/hir-wf-canonicalized.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/wf/hir-wf-canonicalized/hir-wf-canonicalized.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/wf/hir-wf-canonicalized" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/wf/hir-wf-canonicalized/auxiliary"
--- stderr -------------------------------
9
2
9
---
9
2
9
2
error[E0277]: the trait bound `Bar<'a, T>: Foo` is not satisfied
   |
   |
LL |     callback: Box<dyn Callback<dyn Callback<Bar<'a, T>>>>,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Foo` is not implemented for `Bar<'a, T>`

error[E0277]: the trait bound `(dyn Callback<Bar<'a, T>, for<'b, 'c, 'd> Output = ()> + 'static): Foo` is not satisfied
   |
   |
LL |     callback: Box<dyn Callback<dyn Callback<Bar<'a, T>>>>,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Foo` is not implemented for `(dyn Callback<Bar<'a, T>, for<'b, 'c, 'd> Output = ()> + 'static)`
9
2
9
2
---
9
2
9
2
error[E0277]: the size for values of type `(dyn Callback<Bar<'a, T>, for<'b, 'c, 'd> Output = ()> + 'static)` cannot be known at compilation time
   |
   |
LL |     callback: Box<dyn Callback<dyn Callback<Bar<'a, T>>>>,
   |
   |
   = help: the trait `Sized` is not implemented for `(dyn Callback<Bar<'a, T>, for<'b, 'c, 'd> Output = ()> + 'static)`
note: required by a bound in `Bar`
   |
   |
LL | struct Bar<'a, T> {
   |                ^ required by this bound in `Bar`
help: consider relaxing the implicit `Sized` restriction
   |
LL | struct Bar<'a, T: ?Sized> {

9
2
9
```

</details>

--- a/compiler/rustc_middle/src/dep_graph/dep_node.rs
+++ b/compiler/rustc_middle/src/dep_graph/dep_node.rs
@@ -97,7 +97,7 @@ macro_rules! define_dep_nodes {
             // discriminants of the variants have been assigned consecutively from 0
             // so that just the one comparison suffices to check that the u16 can be
             // transmuted to a DepKind.
-            const VARIANTS: u16 = {
+            pub const VARIANTS: u16 = {
                 let deps: &[DepKind] = &[$(DepKind::$variant,)*];
                 let mut i = 0;
                 while i < deps.len() {

--- a/compiler/rustc_middle/src/dep_graph/mod.rs
+++ b/compiler/rustc_middle/src/dep_graph/mod.rs
@@ -26,6 +26,7 @@ pub type DepKindStruct<'tcx> = rustc_query_system::dep_graph::DepKindStruct<TyCt
 impl rustc_query_system::dep_graph::DepKind for DepKind {
     const NULL: Self = DepKind::Null;
     const RED: Self = DepKind::Red;
+    const MAX: u16 = DepKind::VARIANTS - 1;
 
     fn debug_node(node: &DepNode, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         write!(f, "{:?}(", node.kind)?;
@@ -67,6 +68,21 @@ impl rustc_query_system::dep_graph::DepKind for DepKind {
             let Some(icx) = icx else { return };
             op(icx.task_deps)
         })
+    }
+
+    #[track_caller]
+    #[inline]
+    fn from_u16(u: u16) -> Self {
+        if u > Self::MAX {
+            panic!("Invalid DepKind {u}");
+        }
+        // SAFETY: See comment on DepKind::VARIANTS
+        unsafe { std::mem::transmute(u) }
+    }
+
+    #[inline]
+    fn to_u16(self) -> u16 {
+        self as u16
     }
 }
 

--- a/compiler/rustc_query_system/src/dep_graph/edges.rs
+++ b/compiler/rustc_query_system/src/dep_graph/edges.rs
@@ -1,0 +1,60 @@
+use crate::dep_graph::DepNodeIndex;
+use smallvec::SmallVec;
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+
+#[derive(Default, Debug)]
+pub struct EdgesVec {
+    max: u32,
+    edges: SmallVec<[DepNodeIndex; EdgesVec::INLINE_CAPACITY]>,
+}
+
+impl Hash for EdgesVec {
+    #[inline]
+    fn hash<H: Hasher>(&self, hasher: &mut H) {
+        Hash::hash(&self.edges, hasher)
+    }
+}
+
+impl EdgesVec {
+    pub const INLINE_CAPACITY: usize = 8;
+
+    #[inline]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    #[inline]
+    pub fn push(&mut self, edge: DepNodeIndex) {
+        self.max = self.max.max(edge.as_u32());
+        self.edges.push(edge);
+    }
+
+    #[inline]
+    pub fn max_index(&self) -> u32 {
+        self.max
+    }
+}
+
+impl Deref for EdgesVec {
+    type Target = [DepNodeIndex];
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.edges.as_slice()
+    }
+}
+
+impl FromIterator<DepNodeIndex> for EdgesVec {
+    #[inline]
+    fn from_iter<T>(iter: T) -> Self
+    where
+        T: IntoIterator<Item = DepNodeIndex>,
+    {
+        let mut vec = EdgesVec::new();
+        for index in iter {
+            vec.push(index)
+        }
+        vec
+    }
+}

--- a/compiler/rustc_query_system/src/dep_graph/edges.rs
+++ b/compiler/rustc_query_system/src/dep_graph/edges.rs
@@ -1,6 +1,7 @@
 use crate::dep_graph::DepNodeIndex;
 use smallvec::SmallVec;
 use std::hash::{Hash, Hasher};
+use std::iter::Extend;
 use std::ops::Deref;
 
 #[derive(Default, Debug)]
@@ -56,5 +57,17 @@ impl FromIterator<DepNodeIndex> for EdgesVec {
             vec.push(index)
         }
         vec
+    }
+}
+
+impl Extend<DepNodeIndex> for EdgesVec {
+    #[inline]
+    fn extend<T>(&mut self, iter: T)
+    where
+        T: IntoIterator<Item = DepNodeIndex>,
+    {
+        for elem in iter {
+            self.push(elem);
+        }
     }
 }

--- a/compiler/rustc_query_system/src/dep_graph/graph.rs
+++ b/compiler/rustc_query_system/src/dep_graph/graph.rs
@@ -8,7 +8,6 @@ use rustc_data_structures::sync::{AtomicU32, AtomicU64, Lock, Lrc, Ordering};
 use rustc_data_structures::unord::UnordMap;
 use rustc_index::IndexVec;
 use rustc_serialize::opaque::{FileEncodeResult, FileEncoder};
-use smallvec::{smallvec, SmallVec};
 use std::assert_matches::assert_matches;
 use std::collections::hash_map::Entry;
 use std::fmt::Debug;
@@ -19,6 +18,7 @@ use std::sync::atomic::Ordering::Relaxed;
 use super::query::DepGraphQuery;
 use super::serialized::{GraphEncoder, SerializedDepGraph, SerializedDepNodeIndex};
 use super::{DepContext, DepKind, DepNode, HasDepContext, WorkProductId};
+use crate::dep_graph::EdgesVec;
 use crate::ich::StableHashingContext;
 use crate::query::{QueryContext, QuerySideEffects};
 
@@ -137,7 +137,7 @@ impl<K: DepKind> DepGraph<K> {
         let _green_node_index = current.intern_new_node(
             profiler,
             DepNode { kind: DepKind::NULL, hash: current.anon_id_seed.into() },
-            smallvec![],
+            EdgesVec::new(),
             Fingerprint::ZERO,
         );
         assert_eq!(_green_node_index, DepNodeIndex::SINGLETON_DEPENDENCYLESS_ANON_NODE);
@@ -147,7 +147,7 @@ impl<K: DepKind> DepGraph<K> {
             profiler,
             &prev_graph,
             DepNode { kind: DepKind::RED, hash: Fingerprint::ZERO.into() },
-            smallvec![],
+            EdgesVec::new(),
             None,
             false,
         );
@@ -356,12 +356,12 @@ impl<K: DepKind> DepGraphData<K> {
 
         let with_deps = |task_deps| K::with_deps(task_deps, || task(cx, arg));
         let (result, edges) = if cx.dep_context().is_eval_always(key.kind) {
-            (with_deps(TaskDepsRef::EvalAlways), smallvec![])
+            (with_deps(TaskDepsRef::EvalAlways), EdgesVec::new())
         } else {
             let task_deps = Lock::new(TaskDeps {
                 #[cfg(debug_assertions)]
                 node: Some(key),
-                reads: SmallVec::new(),
+                reads: EdgesVec::new(),
                 read_set: Default::default(),
                 phantom_data: PhantomData,
             });
@@ -486,14 +486,14 @@ impl<K: DepKind> DepGraph<K> {
 
                 // As long as we only have a low number of reads we can avoid doing a hash
                 // insert and potentially allocating/reallocating the hashmap
-                let new_read = if task_deps.reads.len() < TASK_DEPS_READS_CAP {
+                let new_read = if task_deps.reads.len() < EdgesVec::INLINE_CAPACITY {
                     task_deps.reads.iter().all(|other| *other != dep_node_index)
                 } else {
                     task_deps.read_set.insert(dep_node_index)
                 };
                 if new_read {
                     task_deps.reads.push(dep_node_index);
-                    if task_deps.reads.len() == TASK_DEPS_READS_CAP {
+                    if task_deps.reads.len() == EdgesVec::INLINE_CAPACITY {
                         // Fill `read_set` with what we have so far so we can use the hashset
                         // next time
                         task_deps.read_set.extend(task_deps.reads.iter().copied());
@@ -572,9 +572,13 @@ impl<K: DepKind> DepGraph<K> {
                 }
             }
 
-            let mut edges = SmallVec::new();
+            let mut edges = EdgesVec::new();
             K::read_deps(|task_deps| match task_deps {
-                TaskDepsRef::Allow(deps) => edges.extend(deps.lock().reads.iter().copied()),
+                TaskDepsRef::Allow(deps) => {
+                    for index in deps.lock().reads.iter().copied() {
+                        edges.push(index);
+                    }
+                }
                 TaskDepsRef::EvalAlways => {
                     edges.push(DepNodeIndex::FOREVER_RED_NODE);
                 }
@@ -872,7 +876,7 @@ impl<K: DepKind> DepGraphData<K> {
 
         let prev_deps = self.previous.edge_targets_from(prev_dep_node_index);
 
-        for &dep_dep_node_index in prev_deps {
+        for dep_dep_node_index in prev_deps {
             self.try_mark_parent_green(qcx, dep_dep_node_index, dep_node, Some(&frame))?;
         }
 
@@ -1308,8 +1312,7 @@ impl<K: DepKind> CurrentDepGraph<K> {
                 let key = prev_graph.index_to_node(prev_index);
                 let edges = prev_graph
                     .edge_targets_from(prev_index)
-                    .iter()
-                    .map(|i| prev_index_to_index[*i].unwrap())
+                    .map(|i| prev_index_to_index[i].unwrap())
                     .collect();
                 let fingerprint = prev_graph.fingerprint_by_index(prev_index);
                 let dep_node_index = self.encoder.borrow().send(profiler, key, fingerprint, edges);
@@ -1334,10 +1337,6 @@ impl<K: DepKind> CurrentDepGraph<K> {
         );
     }
 }
-
-/// The capacity of the `reads` field `SmallVec`
-const TASK_DEPS_READS_CAP: usize = 8;
-type EdgesVec = SmallVec<[DepNodeIndex; TASK_DEPS_READS_CAP]>;
 
 #[derive(Debug, Clone, Copy)]
 pub enum TaskDepsRef<'a, K: DepKind> {

--- a/compiler/rustc_query_system/src/dep_graph/graph.rs
+++ b/compiler/rustc_query_system/src/dep_graph/graph.rs
@@ -574,11 +574,7 @@ impl<K: DepKind> DepGraph<K> {
 
             let mut edges = EdgesVec::new();
             K::read_deps(|task_deps| match task_deps {
-                TaskDepsRef::Allow(deps) => {
-                    for index in deps.lock().reads.iter().copied() {
-                        edges.push(index);
-                    }
-                }
+                TaskDepsRef::Allow(deps) => edges.extend(deps.lock().reads.iter().copied()),
                 TaskDepsRef::EvalAlways => {
                     edges.push(DepNodeIndex::FOREVER_RED_NODE);
                 }

--- a/compiler/rustc_query_system/src/dep_graph/mod.rs
+++ b/compiler/rustc_query_system/src/dep_graph/mod.rs
@@ -1,10 +1,12 @@
 pub mod debug;
 mod dep_node;
+mod edges;
 mod graph;
 mod query;
 mod serialized;
 
 pub use dep_node::{DepKindStruct, DepNode, DepNodeParams, WorkProductId};
+pub use edges::EdgesVec;
 pub use graph::{
     hash_result, DepGraph, DepGraphData, DepNodeColor, DepNodeIndex, TaskDeps, TaskDepsRef,
     WorkProduct, WorkProductMap,
@@ -157,4 +159,10 @@ pub trait DepKind: Copy + fmt::Debug + Eq + Hash + Send + Encodable<FileEncoder>
     fn read_deps<OP>(op: OP)
     where
         OP: for<'a> FnOnce(TaskDepsRef<'a, Self>);
+
+    fn from_u16(u: u16) -> Self;
+
+    fn to_u16(self) -> u16;
+
+    const MAX: u16;
 }

--- a/compiler/rustc_query_system/src/dep_graph/serialized.rs
+++ b/compiler/rustc_query_system/src/dep_graph/serialized.rs
@@ -11,6 +11,29 @@
 //! sequence of NodeInfos to the different arrays in SerializedDepGraph. Since the
 //! node and edge count are stored at the end of the file, all the arrays can be
 //! pre-allocated with the right length.
+//!
+//! The encoding of the de-pgraph is generally designed around the fact that fixed-size
+//! reads of encoded data are generally faster than variable-sized reads. Ergo we adopt
+//! essentially the same varint encoding scheme used in the rmeta format; the edge lists
+//! for each node on the graph store a 2-bit integer which is the number of bytes per edge
+//! index in that node's edge list. We effectively ignore that an edge index of 0 could be
+//! encoded with 0 bytes in order to not require 3 bits to store the byte width of the edges.
+//! The overhead of calculating the correct byte width for each edge is mitigated by
+//! building edge lists with [`EdgesVec`] which keeps a running max of the edges in a node.
+//!
+//! When we decode this data, we do not immediately create [`SerializedDepNodeIndex`] and
+//! instead keep the data in its denser serialized form which lets us turn our on-disk size
+//! efficiency directly into a peak memory reduction. When we convert these encoded-in-memory
+//! values into their fully-deserialized type, we use a fixed-size read of the encoded array
+//! then mask off any errant bytes we read. The array of edge index bytes is padded to permit this.
+//!
+//! We also encode and decode the entire rest of each node using [`SerializedNodeHeader`]
+//! to let this encoding and decoding be done in one fixed-size operation. These headers contain
+//! two [`Fingerprint`]s along with the serialized [`DepKind`], and the number of edge indices
+//! in the node and the number of bytes used to encode the edge indices for this node. The
+//! [`DepKind`], number of edges, and bytes per edge are all bit-packed together, if they fit.
+//! If the number of edges in this node does not fit in the bits available in the header, we
+//! store it directly after the header with leb128.
 
 use super::query::DepGraphQuery;
 use super::{DepKind, DepNode, DepNodeIndex};
@@ -37,7 +60,7 @@ const DEP_NODE_SIZE: usize = std::mem::size_of::<SerializedDepNodeIndex>();
 /// Amount of padding we need to add to the edge list data so that we can retrieve every
 /// SerializedDepNodeIndex with a fixed-size read then mask.
 const DEP_NODE_PAD: usize = DEP_NODE_SIZE - 1;
-/// Amount of bits we need to store the number of used bytes in a SerializedDepNodeIndex.
+/// Number of bits we need to store the number of used bytes in a SerializedDepNodeIndex.
 /// Note that wherever we encode byte widths like this we actually store the number of bytes used
 /// minus 1; for a 4-byte value we technically would have 5 widths to store, but using one byte to
 /// store zeroes (which are relatively rare) is a decent tradeoff to save a bit in our bitfields.
@@ -181,8 +204,15 @@ impl<'a, K: DepKind + Decodable<MemDecoder<'a>>> Decodable<MemDecoder<'a>>
         let mut nodes = IndexVec::with_capacity(node_count);
         let mut fingerprints = IndexVec::with_capacity(node_count);
         let mut edge_list_indices = IndexVec::with_capacity(node_count);
-        // This slightly over-estimates the amount of bytes used for all the edge data but never by
-        // more than ~6%, because over-estimation only occurs for large nodes.
+        // This estimation assumes that all of the encoded bytes are for the edge lists or for the
+        // fixed-size node headers. But that's not necessarily true; if any edge list has a length
+        // that spills out of the size we can bit-pack into SerializedNodeHeader then some of the
+        // total serialized size is also used by leb128-encoded edge list lengths. Neglecting that
+        // contribution to graph_bytes means our estimation of the bytes needed for edge_list_data
+        // slightly overshoots. But it cannot overshoot by much; consider that the worse case is
+        // for a node with length 64, which means the spilled 1-byte leb128 length is 1 byte of at
+        // least (34 byte header + 1 byte len + 64 bytes edge data), which is ~1%. A 2-byte leb128
+        // length is about the same fractional overhead and it amortizes for yet greater lengths.
         let mut edge_list_data = Vec::with_capacity(
             graph_bytes - node_count * std::mem::size_of::<SerializedNodeHeader<K>>(),
         );
@@ -254,10 +284,13 @@ struct Unpacked<K> {
     fingerprint: Fingerprint,
 }
 
-// Bit fields are
-// 0..?    length of the edge
-// ?..?+2  bytes per index
-// ?+2..16 kind
+// Bit fields, where
+// M: bits used to store the length of a node's edge list
+// N: bits used to store the byte width of elements of the edge list
+// are
+// 0..M    length of the edge
+// M..M+N  bytes per index
+// M+N..16 kind
 impl<K: DepKind> SerializedNodeHeader<K> {
     const TOTAL_BITS: usize = std::mem::size_of::<K>() * 8;
     const LEN_BITS: usize = Self::TOTAL_BITS - Self::KIND_BITS - Self::WIDTH_BITS;

--- a/compiler/rustc_query_system/src/dep_graph/serialized.rs
+++ b/compiler/rustc_query_system/src/dep_graph/serialized.rs
@@ -1,6 +1,6 @@
 //! The data that we will serialize and deserialize.
 //!
-//! The dep-graph is serialized as a sequence of NodeInfo, with the dependencies
+//! Notionally, the dep-graph is a sequence of NodeInfo with the dependencies
 //! specified inline. The total number of nodes and edges are stored as the last
 //! 16 bytes of the file, so we can find them easily at decoding time.
 //!
@@ -14,14 +14,16 @@
 
 use super::query::DepGraphQuery;
 use super::{DepKind, DepNode, DepNodeIndex};
+use crate::dep_graph::EdgesVec;
 use rustc_data_structures::fingerprint::Fingerprint;
+use rustc_data_structures::fingerprint::PackedFingerprint;
 use rustc_data_structures::fx::FxHashMap;
 use rustc_data_structures::profiling::SelfProfilerRef;
 use rustc_data_structures::sync::Lock;
 use rustc_index::{Idx, IndexVec};
 use rustc_serialize::opaque::{FileEncodeResult, FileEncoder, IntEncodedWithFixedSize, MemDecoder};
-use rustc_serialize::{Decodable, Decoder, Encodable};
-use smallvec::SmallVec;
+use rustc_serialize::{Decodable, Decoder, Encodable, Encoder};
+use std::marker::PhantomData;
 
 // The maximum value of `SerializedDepNodeIndex` leaves the upper two bits
 // unused so that we can store multiple index types in `CompressedHybridIndex`,
@@ -30,6 +32,16 @@ rustc_index::newtype_index! {
     #[max = 0x7FFF_FFFF]
     pub struct SerializedDepNodeIndex {}
 }
+
+const DEP_NODE_SIZE: usize = std::mem::size_of::<SerializedDepNodeIndex>();
+/// Amount of padding we need to add to the edge list data so that we can retrieve every
+/// SerializedDepNodeIndex with a fixed-size read then mask.
+const DEP_NODE_PAD: usize = DEP_NODE_SIZE - 1;
+/// Amount of bits we need to store the number of used bytes in a SerializedDepNodeIndex.
+/// Note that wherever we encode byte widths like this we actually store the number of bytes used
+/// minus 1; for a 4-byte value we technically would have 5 widths to store, but using one byte to
+/// store zeroes (which are relatively rare) is a decent tradeoff to save a bit in our bitfields.
+const DEP_NODE_WIDTH_BITS: usize = DEP_NODE_SIZE / 2;
 
 /// Data for use when recompiling the **current crate**.
 #[derive(Debug)]
@@ -42,10 +54,10 @@ pub struct SerializedDepGraph<K: DepKind> {
     /// For each DepNode, stores the list of edges originating from that
     /// DepNode. Encoded as a [start, end) pair indexing into edge_list_data,
     /// which holds the actual DepNodeIndices of the target nodes.
-    edge_list_indices: IndexVec<SerializedDepNodeIndex, (u32, u32)>,
-    /// A flattened list of all edge targets in the graph. Edge sources are
-    /// implicit in edge_list_indices.
-    edge_list_data: Vec<SerializedDepNodeIndex>,
+    edge_list_indices: IndexVec<SerializedDepNodeIndex, EdgeHeader>,
+    /// A flattened list of all edge targets in the graph, stored in the same
+    /// varint encoding that we use on disk. Edge sources are implicit in edge_list_indices.
+    edge_list_data: Vec<u8>,
     /// Reciprocal map to `nodes`.
     index: FxHashMap<DepNode<K>, SerializedDepNodeIndex>,
 }
@@ -64,9 +76,35 @@ impl<K: DepKind> Default for SerializedDepGraph<K> {
 
 impl<K: DepKind> SerializedDepGraph<K> {
     #[inline]
-    pub fn edge_targets_from(&self, source: SerializedDepNodeIndex) -> &[SerializedDepNodeIndex] {
-        let targets = self.edge_list_indices[source];
-        &self.edge_list_data[targets.0 as usize..targets.1 as usize]
+    pub fn edge_targets_from(
+        &self,
+        source: SerializedDepNodeIndex,
+    ) -> impl Iterator<Item = SerializedDepNodeIndex> + '_ {
+        let header = self.edge_list_indices[source];
+        let mut raw = &self.edge_list_data[header.start()..];
+        // Figure out where the edge list for `source` ends by getting the start index of the next
+        // edge list, or the end of the array if this is the last edge.
+        let end = self
+            .edge_list_indices
+            .get(source + 1)
+            .map(|h| h.start())
+            .unwrap_or_else(|| self.edge_list_data.len() - DEP_NODE_PAD);
+
+        // The number of edges for this node is implicitly stored in the combination of the byte
+        // width and the length.
+        let bytes_per_index = header.bytes_per_index();
+        let len = (end - header.start()) / bytes_per_index;
+
+        // LLVM doesn't hoist EdgeHeader::mask so we do it ourselves.
+        let mask = header.mask();
+        (0..len).map(move |_| {
+            // Doing this slicing in this order ensures that the first bounds check suffices for
+            // all the others.
+            let index = &raw[..DEP_NODE_SIZE];
+            raw = &raw[bytes_per_index..];
+            let index = u32::from_le_bytes(index.try_into().unwrap()) & mask;
+            SerializedDepNodeIndex::from_u32(index)
+        })
     }
 
     #[inline]
@@ -84,9 +122,40 @@ impl<K: DepKind> SerializedDepGraph<K> {
         self.fingerprints[dep_node_index]
     }
 
+    #[inline]
     pub fn node_count(&self) -> usize {
         self.index.len()
     }
+}
+
+/// A packed representation of an edge's start index and byte width.
+///
+/// This is packed by stealing 2 bits from the start index, which means we only accomodate edge
+/// data arrays up to a quarter of our address space. Which seems fine.
+#[derive(Debug, Clone, Copy)]
+struct EdgeHeader {
+    repr: usize,
+}
+
+impl EdgeHeader {
+    #[inline]
+    fn start(self) -> usize {
+        self.repr >> DEP_NODE_WIDTH_BITS
+    }
+
+    #[inline]
+    fn bytes_per_index(self) -> usize {
+        (self.repr & mask(DEP_NODE_WIDTH_BITS)) + 1
+    }
+
+    #[inline]
+    fn mask(self) -> u32 {
+        mask(self.bytes_per_index() * 8) as u32
+    }
+}
+
+fn mask(bits: usize) -> usize {
+    usize::MAX >> ((std::mem::size_of::<usize>() * 8) - bits)
 }
 
 impl<'a, K: DepKind + Decodable<MemDecoder<'a>>> Decodable<MemDecoder<'a>>
@@ -107,31 +176,50 @@ impl<'a, K: DepKind + Decodable<MemDecoder<'a>>> Decodable<MemDecoder<'a>>
 
         debug!(?node_count, ?edge_count);
 
+        let graph_bytes = d.len() - (2 * IntEncodedWithFixedSize::ENCODED_SIZE) - d.position();
+
         let mut nodes = IndexVec::with_capacity(node_count);
         let mut fingerprints = IndexVec::with_capacity(node_count);
         let mut edge_list_indices = IndexVec::with_capacity(node_count);
-        let mut edge_list_data = Vec::with_capacity(edge_count);
+        // This slightly over-estimates the amount of bytes used for all the edge data but never by
+        // more than ~6%, because over-estimation only occurs for large nodes.
+        let mut edge_list_data = Vec::with_capacity(
+            graph_bytes - node_count * std::mem::size_of::<SerializedNodeHeader<K>>(),
+        );
 
         for _index in 0..node_count {
-            let dep_node: DepNode<K> = Decodable::decode(d);
-            let _i: SerializedDepNodeIndex = nodes.push(dep_node);
+            // Decode the header for this edge; the header packs together as many of the fixed-size
+            // fields as possible to limit the number of times we update decoder state.
+            let node_header = SerializedNodeHeader { bytes: d.read_array(), _marker: PhantomData };
+
+            let _i: SerializedDepNodeIndex = nodes.push(node_header.node());
             debug_assert_eq!(_i.index(), _index);
 
-            let fingerprint: Fingerprint = Decodable::decode(d);
-            let _i: SerializedDepNodeIndex = fingerprints.push(fingerprint);
+            let _i: SerializedDepNodeIndex = fingerprints.push(node_header.fingerprint());
             debug_assert_eq!(_i.index(), _index);
 
-            // Deserialize edges -- sequence of DepNodeIndex
-            let len = d.read_usize();
-            let start = edge_list_data.len().try_into().unwrap();
-            for _ in 0..len {
-                let edge = Decodable::decode(d);
-                edge_list_data.push(edge);
-            }
-            let end = edge_list_data.len().try_into().unwrap();
-            let _i: SerializedDepNodeIndex = edge_list_indices.push((start, end));
+            // If the length of this node's edge list is small, the length is stored in the header.
+            // If it is not, we fall back to another decoder call.
+            let num_edges = node_header.len().unwrap_or_else(|| d.read_usize());
+
+            // The edges index list uses the same varint strategy as rmeta tables; we select the
+            // number of byte elements per-array not per-element. This lets us read the whole edge
+            // list for a node with one decoder call and also use the on-disk format in memory.
+            let edges_len_bytes = node_header.bytes_per_index() * num_edges;
+            // The in-memory structure for the edges list stores the byte width of the edges on
+            // this node with the offset into the global edge data array.
+            let edges_header = node_header.edges_header(&edge_list_data);
+
+            edge_list_data.extend(d.read_raw_bytes(edges_len_bytes));
+
+            let _i: SerializedDepNodeIndex = edge_list_indices.push(edges_header);
             debug_assert_eq!(_i.index(), _index);
         }
+
+        // When we access the edge list data, we do a fixed-size read from the edge list data then
+        // mask off the bytes that aren't for that edge index, so the last read may dangle off the
+        // end of the array. This padding ensure it doesn't.
+        edge_list_data.extend(&[0u8; DEP_NODE_PAD]);
 
         let index: FxHashMap<_, _> =
             nodes.iter_enumerated().map(|(idx, &dep_node)| (dep_node, idx)).collect();
@@ -140,11 +228,151 @@ impl<'a, K: DepKind + Decodable<MemDecoder<'a>>> Decodable<MemDecoder<'a>>
     }
 }
 
-#[derive(Debug, Encodable, Decodable)]
-pub struct NodeInfo<K: DepKind> {
+/// A packed representation of all the fixed-size fields in a `NodeInfo`.
+///
+/// This stores in one byte array:
+/// * The `Fingerprint` in the `NodeInfo`
+/// * The `Fingerprint` in `DepNode` that is in this `NodeInfo`
+/// * The `DepKind`'s discriminant (a u16, but not all bits are used...)
+/// * The byte width of the encoded edges for this node
+/// * In whatever bits remain, the length of the edge list for this node, if it fits
+struct SerializedNodeHeader<K> {
+    // 2 bytes for the DepNode
+    // 16 for Fingerprint in DepNode
+    // 16 for Fingerprint in NodeInfo
+    bytes: [u8; 34],
+    _marker: PhantomData<K>,
+}
+
+// The fields of a `SerializedNodeHeader`, this struct is an implementation detail and exists only
+// to make the implementation of `SerializedNodeHeader` simpler.
+struct Unpacked<K> {
+    len: Option<usize>,
+    bytes_per_index: usize,
+    kind: K,
+    hash: PackedFingerprint,
+    fingerprint: Fingerprint,
+}
+
+// Bit fields are
+// 0..?    length of the edge
+// ?..?+2  bytes per index
+// ?+2..16 kind
+impl<K: DepKind> SerializedNodeHeader<K> {
+    const TOTAL_BITS: usize = std::mem::size_of::<K>() * 8;
+    const LEN_BITS: usize = Self::TOTAL_BITS - Self::KIND_BITS - Self::WIDTH_BITS;
+    const WIDTH_BITS: usize = DEP_NODE_WIDTH_BITS;
+    const KIND_BITS: usize = Self::TOTAL_BITS - K::MAX.leading_zeros() as usize;
+    const MAX_INLINE_LEN: usize = (u16::MAX as usize >> (Self::TOTAL_BITS - Self::LEN_BITS)) - 1;
+
+    #[inline]
+    fn new(node_info: &NodeInfo<K>) -> Self {
+        debug_assert_eq!(Self::TOTAL_BITS, Self::LEN_BITS + Self::WIDTH_BITS + Self::KIND_BITS);
+
+        let NodeInfo { node, fingerprint, edges } = node_info;
+
+        let mut head = node.kind.to_u16();
+
+        let free_bytes = edges.max_index().leading_zeros() as usize / 8;
+        let bytes_per_index = (DEP_NODE_SIZE - free_bytes).saturating_sub(1);
+        head |= (bytes_per_index as u16) << Self::KIND_BITS;
+
+        // Encode number of edges + 1 so that we can reserve 0 to indicate that the len doesn't fit
+        // in this bitfield.
+        if edges.len() <= Self::MAX_INLINE_LEN {
+            head |= (edges.len() as u16 + 1) << (Self::KIND_BITS + Self::WIDTH_BITS);
+        }
+
+        let hash: Fingerprint = node.hash.into();
+
+        // Using half-open ranges ensures an unconditional panic if we get the magic numbers wrong.
+        let mut bytes = [0u8; 34];
+        bytes[..2].copy_from_slice(&head.to_le_bytes());
+        bytes[2..18].copy_from_slice(&hash.to_le_bytes());
+        bytes[18..].copy_from_slice(&fingerprint.to_le_bytes());
+
+        #[cfg(debug_assertions)]
+        {
+            let res = Self { bytes, _marker: PhantomData };
+            assert_eq!(node_info.fingerprint, res.fingerprint());
+            assert_eq!(node_info.node, res.node());
+            if let Some(len) = res.len() {
+                assert_eq!(node_info.edges.len(), len);
+            }
+        }
+        Self { bytes, _marker: PhantomData }
+    }
+
+    #[inline]
+    fn unpack(&self) -> Unpacked<K> {
+        let head = u16::from_le_bytes(self.bytes[..2].try_into().unwrap());
+        let hash = self.bytes[2..18].try_into().unwrap();
+        let fingerprint = self.bytes[18..].try_into().unwrap();
+
+        let kind = head & mask(Self::KIND_BITS) as u16;
+        let bytes_per_index = (head >> Self::KIND_BITS) & mask(Self::WIDTH_BITS) as u16;
+        let len = (head as usize) >> (Self::WIDTH_BITS + Self::KIND_BITS);
+
+        Unpacked {
+            len: len.checked_sub(1),
+            bytes_per_index: bytes_per_index as usize + 1,
+            kind: DepKind::from_u16(kind),
+            hash: Fingerprint::from_le_bytes(hash).into(),
+            fingerprint: Fingerprint::from_le_bytes(fingerprint),
+        }
+    }
+
+    #[inline]
+    fn len(&self) -> Option<usize> {
+        self.unpack().len
+    }
+
+    #[inline]
+    fn bytes_per_index(&self) -> usize {
+        self.unpack().bytes_per_index
+    }
+
+    #[inline]
+    fn fingerprint(&self) -> Fingerprint {
+        self.unpack().fingerprint
+    }
+
+    #[inline]
+    fn node(&self) -> DepNode<K> {
+        let Unpacked { kind, hash, .. } = self.unpack();
+        DepNode { kind, hash }
+    }
+
+    #[inline]
+    fn edges_header(&self, edge_list_data: &[u8]) -> EdgeHeader {
+        EdgeHeader {
+            repr: (edge_list_data.len() << DEP_NODE_WIDTH_BITS) | (self.bytes_per_index() - 1),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct NodeInfo<K: DepKind> {
     node: DepNode<K>,
     fingerprint: Fingerprint,
-    edges: SmallVec<[DepNodeIndex; 8]>,
+    edges: EdgesVec,
+}
+
+impl<K: DepKind> Encodable<FileEncoder> for NodeInfo<K> {
+    fn encode(&self, e: &mut FileEncoder) {
+        let header = SerializedNodeHeader::new(self);
+        e.emit_raw_bytes(&header.bytes);
+
+        if header.len().is_none() {
+            e.emit_usize(self.edges.len());
+        }
+
+        let bytes_per_index = header.bytes_per_index();
+        for node_index in self.edges.iter() {
+            let bytes = node_index.as_u32().to_le_bytes();
+            e.emit_raw_bytes(&bytes[..bytes_per_index]);
+        }
+    }
 }
 
 struct Stat<K: DepKind> {
@@ -303,7 +531,7 @@ impl<K: DepKind + Encodable<FileEncoder>> GraphEncoder<K> {
         profiler: &SelfProfilerRef,
         node: DepNode<K>,
         fingerprint: Fingerprint,
-        edges: SmallVec<[DepNodeIndex; 8]>,
+        edges: EdgesVec,
     ) -> DepNodeIndex {
         let _prof_timer = profiler.generic_activity("incr_comp_encode_dep_graph");
         let node = NodeInfo { node, fingerprint, edges };

--- a/compiler/rustc_query_system/src/lib.rs
+++ b/compiler/rustc_query_system/src/lib.rs
@@ -4,6 +4,7 @@
 #![feature(min_specialization)]
 #![feature(extern_types)]
 #![feature(let_chains)]
+#![feature(inline_const)]
 #![allow(rustc::potential_query_instability)]
 #![deny(rustc::untranslatable_diagnostic)]
 #![deny(rustc::diagnostic_outside_of_impl)]

--- a/compiler/rustc_serialize/src/opaque.rs
+++ b/compiler/rustc_serialize/src/opaque.rs
@@ -353,7 +353,7 @@ impl<'a> MemDecoder<'a> {
     }
 
     #[inline]
-    fn read_array<const N: usize>(&mut self) -> [u8; N] {
+    pub fn read_array<const N: usize>(&mut self) -> [u8; N] {
         self.read_raw_bytes(N).try_into().unwrap()
     }
 

